pegen/pegen.c

```diff
@@ -1677,6 +1615,973 @@ parsestr(Parser *p, const char *s, int *bytesmode, int *rawmode, PyObject **resu
+   expression, in order to allow leading spaces in the expression. */
+static expr_ty
+fstring_compile_expr(Parser *p, const char *expr_start, const char *expr_end)
+{
+    mod_ty mod;
+    char *str;
+    Py_ssize_t len;
+    const char *s;
+
+    assert(expr_end >= expr_start);
+    assert(*(expr_start-1) == '{');
+    assert(*expr_end == '}' || *expr_end == '!' || *expr_end == ':' ||
+           *expr_end == '=');
+
+    /* If the substring is all whitespace, it's an error.  We need to catch this
+       here, and not when we call PyParser_SimpleParseStringFlagsFilename,
+       because turning the expression '' in to '()' would go from being invalid
+       to valid. */
+    for (s = expr_start; s != expr_end; s++) {
+        char c = *s;
+        /* The Python parser ignores only the following whitespace
+           characters (\r already is converted to \n). */
+        if (!(c == ' ' || c == '\t' || c == '\n' || c == '\f')) {
+            break;
+        }
+    }
+    if (s == expr_end) {
+        raise_syntax_error(p, "f-string: empty expression not allowed");
+        return NULL;
+    }
+
+    len = expr_end - expr_start;
+    /* Allocate 3 extra bytes: open paren, close paren, null byte. */
+    str = PyMem_RawMalloc(len + 3);
+    if (str == NULL) {
+        PyErr_NoMemory();
+        return NULL;
+    }
+
+    str[0] = '(';
+    memcpy(str+1, expr_start, len);
+    str[len+1] = ')';
+    str[len+2] = 0;
+
+    struct tok_state* tok = PyTokenizer_FromString(str, 1);
+    if (tok == NULL) {
+        return NULL;
+    }
+    mod_ty (*the_start_rule)(Parser*) = p->start_rule_func;
+
+    Parser *p2 = PyMem_Malloc(sizeof(Parser));
+    if (p2 == NULL) {
+        PyErr_Format(PyExc_MemoryError, "Out of memory for Parser");
+        goto exit;
+    }
+    p2->tok = tok;
+    p2->keywords = p->keywords;
+    p2->n_keyword_lists = p->n_keyword_lists;
+    p2->tokens = PyMem_Malloc(sizeof(Token *));
+    if (!p2->tokens) {
+        PyErr_Format(PyExc_MemoryError, "Out of memory for tokens");
+        goto exit;
+    }
+    p2->tokens[0] = PyMem_Malloc(sizeof(Token));
+    memset(p2->tokens[0], '\0', sizeof(Token));
+    p2->mark = 0;
+    p2->fill = 0;
+    p2->size = 1;
+    p2->arena = p->arena;
+    if (fill_token(p2) < 0) {
+        goto exit;
+    }
+    PyErr_Clear();
+    mod = the_start_rule(p2);
+
+
+    PyTokenizer_Free(tok);
+
+    /* Reuse str to find the correct column offset. */
+    str[0] = '{';
+    str[len+1] = '}';
+    if (!mod) {
+        return NULL;
+    }
+
+exit:
+    for (int i = 0; i < p2->size; i++) {
+        PyMem_Free(p2->tokens[i]);
+    }
+    PyMem_Free(p2->tokens);
+    PyMem_Free(p2);
+    if (mod == NULL) {
+        return NULL;
+    }
+    if (asdl_seq_LEN(mod->v.Module.body) != 1) {
+        raise_syntax_error(p, "f-string: invalid expression");
+    }
+    expr_ty expr = asdl_seq_GET(mod->v.Module.body, 0);
+    return expr->v.Subscript.value;
```

Comment:
Why are you using `Subscript` here?

Reply:
Because for some reason that I am currently investigating, the first  (and only) expression in the `mod` is enclosed in a `Subscript`:

```
(gdb) p mod
$4 = (mod_ty) 0x555555d6ff78
(gdb) p mod.kind 
$5 = Module_kind
(gdb) p expr
$6 = (expr_ty) 0x555555d6f5d0
(gdb) p expr.kind
$7 = Subscript_kind
```

help to understand what is going on here would be appreciated ;)

Reply:
Oh, is because `asdl_seq_GET(mod->v.Module.body, 0)` should be `stmt_ty` not `expr_ty`. Pushing a fix